runtime/intern.c

```diff
@@ -35,38 +35,87 @@
 #include "caml/mlvalues.h"
 #include "caml/misc.h"
 #include "caml/reverse.h"
+#include "caml/shared_heap.h"
 #include "caml/signals.h"
 
+/* Item on the stack with defined operation */
+struct intern_item {
+  value * dest;
+  intnat arg;
+  enum {
+    OReadItems, /* read arg items and store them in dest[0], dest[1], ... */
+    OFreshOID,  /* generate a fresh OID and store it in *dest */
+    OShift      /* offset *dest by arg */
+  } op;
+};
+
+/* FIXME: This is duplicated in two other places, with the only difference of
+   the type of elements stored in the stack. Possible solution in C would
+   be to instantiate stack these function via. C preprocessor macro.
+ */
+
+#define INTERN_STACK_INIT_SIZE 256
+#define INTERN_STACK_MAX_SIZE (1024*1024*100)
 
-static unsigned char * intern_src;
-/* Reading pointer in block holding input data. */
+struct caml_intern_state {
 
-static unsigned char * intern_input = NULL;
-/* Pointer to beginning of block holding input data,
-   if non-NULL this pointer will be freed by the cleanup function. */
+  const unsigned char * intern_src;
+  /* Reading pointer in block holding input data. */
 
-static header_t * intern_dest;
-/* Writing pointer in destination block */
+  unsigned char * intern_input;
+  /* Pointer to beginning of block holding input data,
+    if non-NULL this pointer will be freed by the cleanup function. */
 
-static char * intern_extra_block = NULL;
-/* If non-NULL, point to new heap chunk allocated with caml_alloc_for_heap. */
+  asize_t obj_counter;
+  /* Count how many objects seen so far */
 
-static asize_t obj_counter;
-/* Count how many objects seen so far */
+  value * intern_obj_table;
+  /* The pointers to objects already seen */
 
-static value * intern_obj_table = NULL;
-/* The pointers to objects already seen */
+  struct intern_item intern_stack_init[INTERN_STACK_INIT_SIZE];
+  /* The initial intern stack */
 
-static color_t intern_color;
-/* Color to assign to newly created headers */
+  struct intern_item * intern_stack;
+  /* Initially points to [intern_stack_init] */
 
-static header_t intern_header;
-/* Original header of the destination block.
-   Meaningful only if intern_extra_block is NULL. */
+  struct intern_item * intern_stack_limit;
 
-static value intern_block = 0;
-/* Point to the heap block allocated as destination block.
-   Meaningful only if intern_extra_block is NULL. */
+  header_t * intern_dest;
+  /* Writing pointer in destination block. Only used when the object fits in
+     the minor heap. */
+};
+
+/* Allocates the domain local intern state if needed */
+static struct caml_intern_state* get_intern_state ()
+{
+  struct caml_intern_state* s;
+
+  if (Caml_state->intern_state != NULL)
+    return Caml_state->intern_state;
+
+  s = caml_stat_alloc_noexc(sizeof(struct caml_intern_state));
+  if (s == NULL) {
+    caml_raise_out_of_memory ();
+  }
```

Comment:
Could be simplified:
```suggestion
  s = caml_stat_alloc(sizeof(struct caml_intern_state));
```